src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/MoreTimeRequestedCallbackService.java

```diff
@@ -0,0 +1,116 @@
+import uk.gov.hmcts.cmc.domain.models.Claim;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse.AboutToStartOrSubmitCallbackResponseBuilder;
+import uk.gov.hmcts.reform.ccd.client.model.CallbackRequest;
+import uk.gov.hmcts.reform.ccd.client.model.CallbackResponse;
+import uk.gov.hmcts.reform.ccd.client.model.SubmittedCallbackResponse;
+
+import java.time.LocalDate;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.lang.String.format;
+import static uk.gov.hmcts.cmc.claimstore.appinsights.AppInsights.REFERENCE_NUMBER;
+import static uk.gov.hmcts.cmc.claimstore.appinsights.AppInsightsEvent.RESPONSE_MORE_TIME_REQUESTED_PAPER;
+
+@Service
+public class MoreTimeRequestedCallbackService {
+
+    private final EventProducer eventProducer;
+    private final AppInsights appInsights;
+    private final ResponseDeadlineCalculator responseDeadlineCalculator;
+    private final MoreTimeRequestRule moreTimeRequestRule;
+    private final CCDCaseDataToClaim ccdCaseDataToClaim;
+
+    @Autowired
+    public MoreTimeRequestedCallbackService(
+        EventProducer eventProducer,
+        AppInsights appInsights,
+        ResponseDeadlineCalculator responseDeadlineCalculator,
+        MoreTimeRequestRule moreTimeRequestRule,
+        CCDCaseDataToClaim ccdCaseDataToClaim) {
+        this.eventProducer = eventProducer;
+        this.appInsights = appInsights;
+        this.responseDeadlineCalculator = responseDeadlineCalculator;
+        this.moreTimeRequestRule = moreTimeRequestRule;
+        this.ccdCaseDataToClaim = ccdCaseDataToClaim;
+    }
+
+    public CallbackResponse execute(CallbackType callbackType, CallbackRequest callbackRequest) {
+        switch (callbackType) {
+            case ABOUT_TO_START:
+                return requestMoreTimeOnPaper(callbackRequest, true);
+            case ABOUT_TO_SUBMIT:
+                return requestMoreTimeOnPaper(callbackRequest, false);
+            case SUBMITTED:
+                return requestMoreTimeOnPaperSubmitted(callbackRequest);
+            default:
+                throw new IllegalArgumentException(
+                    format("Callback for event %s, type %s not implemented",
+                        callbackRequest.getEventId(),
+                        callbackType));
+        }
+    }
+
+    private AboutToStartOrSubmitCallbackResponse requestMoreTimeOnPaper(
+        CallbackRequest callbackRequest,
+        boolean validateOnly
+    ) {
+        Claim claim = convertCallbackToClaim(callbackRequest);
+
+        List<String> validationResult = moreTimeRequestRule.validateMoreTimeCanBeRequested(claim);
+        AboutToStartOrSubmitCallbackResponseBuilder builder = AboutToStartOrSubmitCallbackResponse
+            .builder();
+
+        if (validateOnly || !validationResult.isEmpty()) {
```

Comment:
Rather than controlling based on the boolean parameter, it would be nice to split this method in two and name accordingly.

Reply:
I didn't write that code, I just moved it out to a separate service. I'd rather not change the logic of another, unrelated service as part of this PR, as it might be a can of worm :D but up to you

Reply:
Please do refactor. :) It's very simple one.

Reply:
👍